src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -209,37 +209,44 @@ ss::future<init_tm_tx_reply> tx_gateway_frontend::dispatch_init_tm_tx(
       });
 }
 
-ss::future<cluster::init_tm_tx_reply> tx_gateway_frontend::do_init_tm_tx(
+ss::future<init_tm_tx_reply> tx_gateway_frontend::do_init_tm_tx(
   ss::shard_id shard,
   kafka::transactional_id tx_id,
   model::timeout_clock::duration timeout) {
     return container().invoke_on(
       shard, _ssg, [tx_id, timeout](tx_gateway_frontend& self) {
-          return self.do_init_tm_tx(tx_id, timeout);
-      });
-}
+          auto partition = self._partition_manager.local().get(
+            model::tx_manager_ntp);
+          if (!partition) {
+              vlog(
+                clusterlog.warn,
+                "can't get partition by {} ntp",
+                model::tx_manager_ntp);
+              return ss::make_ready_future<init_tm_tx_reply>(
+                init_tm_tx_reply{.ec = tx_errc::partition_not_found});
+          }
 
-ss::future<cluster::init_tm_tx_reply> tx_gateway_frontend::do_init_tm_tx(
-  kafka::transactional_id tx_id, model::timeout_clock::duration timeout) {
-    auto partition = _partition_manager.local().get(model::tx_manager_ntp);
-    if (!partition) {
-        vlog(
-          clusterlog.warn,
-          "can't get partition by {} ntp",
-          model::tx_manager_ntp);
-        co_return cluster::init_tm_tx_reply{.ec = tx_errc::partition_not_found};
-    }
+          auto stm = partition->tm_stm();
 
-    auto stm = partition->tm_stm();
+          if (!stm) {
+              vlog(
+                clusterlog.warn,
+                "can't get tm stm of the {}' partition",
+                model::tx_manager_ntp);
+              return ss::make_ready_future<init_tm_tx_reply>(
```

Comment:
Maybe we should crash in this scenario with `vassert` ?

Reply:
I'm leaning towards graceful degradation. The situation in-question is misconfiguration of the transaction layer: a request handler thinks transactions are on but a partition thinks they are off. If use vassert all the customer's workloads become broken but here we refuse to process only the transactional workloads so Redpanda remains useful for at least for some kind of requests.